src/ripple/rpc/impl/ShardArchiveHandler.cpp

```diff
@@ -154,20 +315,28 @@ ShardArchiveHandler::next(std::lock_guard<std::mutex>& l)
         return next(l);
     }
 
-    // Download the archive
+    // Download the archive. Process in another thread
+    // to prevent holding up the lock if the downloader
+    // sleeps.
     auto const& url {archives_.begin()->second};
-    if (!downloader_->download(
-        url.domain,
-        std::to_string(url.port.get_value_or(443)),
-        url.path,
-        11,
-        dstDir / "archive.tar.lz4",
-        std::bind(&ShardArchiveHandler::complete,
-            shared_from_this(), std::placeholders::_1)))
-    {
-        remove(l);
-        return next(l);
-    }
+    app_.getJobQueue().addJob(
```

Comment:
I already take care of the deferred execution problem in `ShardArchiveHandler::onStop()` by waiting on a condition variable in `SSLHTTPDownloader`. When I return from `ShardArchiveHandler::onStop()` I know that the download has been cancelled, and shutdown can proceed. The `ClosureCounter` only provides reference counting, which isn't necessary since there is at most only one deferred job. I need to actively cancel the download, and using the `ClosureCounter` and simply waiting for the download to finish while the node is shutting down isn't acceptable.

Reply:
If you're sure you have the lifetime stuff all under control, I'm good with your analysis and testing.  I'm also more comfortable now that you're going to make a run at turning the `ShardArchveHandler` from a `shared_ptr` singleton into an ordinary member of `Application`.  I think that will be a big improvement in lifetime management.  :+1:

Reply:
I've decided to use the `ClosureCounter` to provide better coverage against the risk of in-flight handlers still running during shutdown and made the `ShardArchiveHandler` a single `unique_ptr` held by the `Application` which is the standing convention.